### PR TITLE
feat: introduce Form Group Header to Form Group

### PR DESCRIPTION
**Pull request by kavya-b**
## Related Issue
Closes SAP/fundamental-styles#1880

## Description
PR introduces Form Group Header to Form Group

### Before:
N/A

### After:
Screenshot 2020-11-12 at 10 04 42 PM
Screenshot 2020-11-12 at 10 04 56 PM

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [n/a] hover state of the element follow design spec
- [n/a] focus state of the element follow design spec
- [n/a] active state of the element follow design spec
- [n/a] selected state of the element follow design spec
- [n/a] selected hover state of the element follow design spec
- [n/a] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [x] only one top level `fd-*` class is used in the file
- [x] BEM naming convention is used
- [x] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [x] `fd-reset()` mixin is applied to all elements
- [x] Variables are used, if some value is used more than twice.
- [x] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] tested Storybook examples with "CSS Resources" `normalize` option 
- [x] tested Storybook examples with "CSS Resources" `unnormalize` option 
- [x] Verified all styles in IE11
- [x] Updated tests
4. Documentation
- [x] Storybook documentation has been created/updated
- [x] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

### CODE REVIEW

**Comment by kavya-b** [comment]
I have tried running `npm run test` and `npm run test:update` but strangely the storyshots don't update with the newly added code. I get only empty blocks like this:
image
Therefore I haven't added the storyshot file here and travis will fail until I fix this.
Am I doing something wrong, @InnaAtanasova @prsdthkr ?

File: stories/form-group/form-group.stories.js [hunk 1]
```diff
@@ -36,3 +36,125 @@ export const required = () => `<div class="fd-form-group">
         </div>
     </div>
 `;
+
+export const groupHeader = () => `
+    <div class="fd-form-group">
+        <div class="fd-form-group__header">
+            <span class="fd-form-group__header-text">Group Header</span>
```
Comment (meganaconley) [inline stories/form-group/form-group.stories.js:43]: I see you checked off the a11y requirements on the PR template, what did you test this with? I'm not able to get this read out when accessing the form with VoiceOver. I think it needs to be associated using `aria-labelledby`. See Approach 2 here for an example:

https://www.w3.org/WAI/tutorials/forms/grouping/

**Review by meganaconley — CHANGES_REQUESTED** [review]

File: stories/form-group/form-group.stories.js (hunk 1 above)
Comment (kavya-b) [reply stories/form-group/form-group.stories.js:43]: pressing Ctrl+Option+Shift+Right Arrow read out the header text for me in VoiceOver.

File: stories/form-group/form-group.stories.js [hunk 2]
```diff
@@ -36,3 +36,125 @@ export const required = () => `<div class="fd-form-group">
...
+                <label class="fd-form-label" for="input-14c-name">Name:</label>
+                </div>
+                <div class="fd-col fd-col-md--10 fd-col-lg--8 fd-col-xl--12">
+                <input class="fd-input" type="text" id="input-14c-name" placeholder="Enter First and Last Name" value="Amelia Perry">
+                </div>
+            </div>
+        </div>
+        <div class="fd-form-group fd-col fd-col-md--6 fd-col-lg--6 fd-col-xl--6 fd-col--wrap">
+            <div class="fd-form-group__header">
+                <span class="fd-form-group__header-text">Group Header 3</span>
```
Comment (salarenko) [inline stories/form-group/form-group.stories.js:135]: Shouldn't he a `h` tag?

File: stories/form-group/form-group.stories.js [hunk 3]
```diff
@@ -36,3 +36,125 @@ export const required = () => `<div class="fd-form-group">
...
+            </div>
+            <div class="fd-row">
+                <div class="fd-col fd-col-md--2 fd-col-lg--4 fd-col-xl--12">
+                <label class="fd-form-label" for="input-14j-name">Name:</label>
+                </div>
+                <div class="fd-col fd-col-md--10 fd-col-lg--8 fd-col-xl--12">
+                <input class="fd-input" type="text" id="input-14j-name" placeholder="Enter First and Last Name" value="Amelia Perry">
+                </div>
+                <div class="fd-col fd-col-md--2 fd-col-lg--4 fd-col-xl--12">
+                <label class="fd-form-label" for="input-14h-name">Name:</label>
```
Comment (salarenko) [inline stories/form-group/form-group.stories.js:145]: I'm not sure if keeping label and input in separate grid columns is a good idea. You need to make double work to maintain responsive grid. In my opinion this is quite unusual approach bringing to value but more code into the HTML.

File: stories/form-group/form-group.stories.js [hunk 4]
```diff
@@ -36,3 +36,125 @@ export const required = () => `<div class="fd-form-group">
...
+                </div>
+            </div>
+        </div>
+
+        <div class="fd-form-group fd-col fd-col-md--6 fd-col-lg--6 fd-col-xl--6 fd-col--wrap">
+            <div class="fd-form-group__header">
+                <span class="fd-form-group__header-text">Group Header 2</span>
+            </div>
+            <div class="fd-row">
+                <div class="fd-col fd-col-md--2 fd-col-lg--4 fd-col-xl--12">
```
Comment (salarenko) [inline stories/form-group/form-group.stories.js:113]: This examples should use `form-item`.

**Review by salarenko — COMMENTED** [review]
- [x] Check A11y violations (mostly non-unique id issues)
- [ ] We don't need separate RTL examples, user can switch the direction using documentation built in control.

**Review by salarenko — CHANGES_REQUESTED** [review]
Needs changes

File: stories/form-group/form-group.stories.js [hunk 5]
```diff
@@ -36,3 +36,125 @@ export const required = () => `<div class="fd-form-group">
...
+                </div>
+            </div>
+        </div>
+        <div class="fd-form-group fd-col fd-col-md--6 fd-col-lg--6 fd-col-xl--6 fd-col--wrap">
+            <div class="fd-form-group__header">
+                <span class="fd-form-group__header-text">Group Header 3</span>
+            </div>
+            <div class="fd-row">
+                <div class="fd-col fd-col-md--2 fd-col-lg--4 fd-col-xl--12">
+                <label class="fd-form-label" for="input-14j-name">Name:</label>
```
[the remark this answers is not in the document]
Comment (kavya-b) [reply stories/form-group/form-group.stories.js]: fixed

File: stories/form-group/form-group.stories.js (hunk 2 above)
Comment (kavya-b) [reply stories/form-group/form-group.stories.js]: replaced with `h1` tags

File: stories/form-group/form-group.stories.js (hunk 3 above)
Comment (kavya-b) [reply stories/form-group/form-group.stories.js]: I am just following the existing Form Grid example here and have not modified any underlying logic for responsive grid column usage.

File: stories/form-group/form-group.stories.js (hunk 4 above)
Comment (kavya-b) [reply stories/form-group/form-group.stories.js]: using `fd-form-item` now

File: stories/form-group/form-group.stories.js (hunk 1 above)
Comment (kavya-b) [reply stories/form-group/form-group.stories.js]: added `aria-labelledby` with `id` and `role=group` to get VoiceOver to read the group name on entering the group.

File: src/form-layout-grid.scss [hunk 6]
```diff
@@ -85,6 +91,13 @@ $cols: 11;
           margin: 0 -0.25rem;
         }
       }
+
+      .#{$fd-namespace}-form-group.#{$blockCol} {
+        padding: 1rem;
+      }
+      .#{$fd-namespace}-form-item.#{$blockRow} {
+        flex-direction: row;
+      }
```
Comment (JKMarkowski) [inline src/form-layout-grid.scss:100]: Some time ago we decided to not mix blocks in files. Is there any way. Could you create other, stronger element to overwrite `.#{$fd-namespace}-form-group` / `.#{$fd-namespace}-form-item`. Or use them in different elements?

Comment (JKMarkowski) [reply src/form-layout-grid.scss:100]: image

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
Check my previous comment, let me know if something is not clear

File: src/form-layout-grid.scss (hunk 6 above)
Comment (kavya-b) [reply src/form-layout-grid.scss:100]: so would it be okay to have separate element modifiers such as this?:
image

**Comment by kavya-b** [comment]
@JKMarkowski  using `fd-row__form-item` and `fd-col__form-group` elements now, please check.

**Review by meganaconley — APPROVED** [review]

File: src/form-layout-grid.scss (hunk 6 above)
Comment (JKMarkowski) [reply src/form-layout-grid.scss]: Yes, it's much better there

**Review by JKMarkowski — APPROVED** [review]
LGTM

**Review by InnaAtanasova — APPROVED** [review]

**Comment by kavya-b** [comment]
@salarenko could you please check this again, your comments were addressed; thanks!

**Review by salarenko — APPROVED** [review]

**Pull request merged**